guppylang/hresult.py

```diff
@@ -0,0 +1,206 @@
+    ):
+        self.results = [
+            res if isinstance(res, HResult) else HResult(res) for res in results or []
+        ]
+
+    def register_counts(
+        self, strict_names: bool = False, strict_lengths: bool = False
+    ) -> dict[str, Counter[str]]:
+        """Convert results to a dictionary of register counts.
+
+        Returns:
+            dict: A dictionary where the keys are the register names
+            and the values are the counts of the register bitstrings.
+        """
+        return {
+            reg: Counter(bitstrs)
+            for reg, bitstrs in self.register_bitstrings(
+                strict_lengths=strict_lengths, strict_names=strict_names
+            ).items()
+        }
+
+    def register_bitstrings(
+        self, strict_names: bool = False, strict_lengths: bool = False
+    ) -> dict[str, list[str]]:
+        """Convert results to a dictionary from register name to list of bitstrings over
+        the shots.
+
+        Args:
+            strict_names: Whether to enforce that all shots have the same
+                registers.
+            strict_lengths: Whether to enforce that all register bitstrings have
+                the same length.
+
+        """
+
+        shot_dct: dict[str, list[str]] = defaultdict(list)
+        for shot in self.results:
+            bitstrs = shot.to_register_bits()
+            for reg, bitstr in bitstrs.items():
+                if (
+                    strict_lengths
+                    and reg in shot_dct
+                    and len(shot_dct[reg][0]) != len(bitstr)
+                ):
+                    raise ValueError(
+                        "All register bitstrings must have the same length."
+                    )
+                shot_dct[reg].append(bitstr)
+            if strict_names and not bitstrs.keys() == shot_dct.keys():
+                raise ValueError("All shots must have the same registers.")
+        return shot_dct
+
+    def to_pytket(self) -> BackendResult:
+        """Convert results to a pytket BackendResult.
+
+        Returns:
+            BackendResult: A BackendResult object with the shots.
+
+        Raises:
+            ImportError: If pytket is not installed.
+            ValueError: If a register's bitstrings have different lengths or not all
+            registers are present in all shots.
+        """
+        try:
+            from pytket._tket.unit_id import Bit
+            from pytket.backends.backendresult import BackendResult
+            from pytket.utils.outcomearray import OutcomeArray
+        except ImportError as e:
+            raise ImportError(
+                "Pytket is an optional dependency, install with the `pytket` extra"
+            ) from e
+        counts = self.register_bitstrings(strict_lengths=True, strict_names=True)
+        reg_sizes: dict[str, int] = {
+            reg: len(next(iter(counts[reg]), "")) for reg in counts
```

Comment:
Why is the `""` default in `next` needed? Shouldn't all registers be populated with at least one value?

Reply:
yes, it is just there to avoid having to handle unexpected stopiteration errors induced by changes to other code. It is safe to assume unpopulated registers are size 0